src/main/cpp/subsystems/SwerveDrive.cpp

```diff
@@ -0,0 +1,559 @@
+    } else if (auto_state == TRANSLATE) {
+        // if we're close enough to the goal then we probably don't have to drive and this was just to turn
+        double error = 0.0;
+        OKC_CALL(this->left_front_module->GetDriveError(&error));
+        if (abs(distance_to_goal) < 0.1 || abs(error) < 0.1) {
+            auto_state = COMPLETE;
+
+            // stop the drive motors then
+            this->interface_->left_front_drive_motor_output = 0.0;
+            this->interface_->left_back_drive_motor_output = 0.0;
+            this->interface_->right_front_drive_motor_output = 0.0;
+            this->interface_->right_back_drive_motor_output = 0.0;
+            return true;
+        }
+
+        // drive_pid the distance
+        OKC_CALL(this->left_front_module->SetDistance(distance_to_goal));
+        OKC_CALL(this->left_back_module->SetDistance(distance_to_goal));
+        OKC_CALL(this->right_front_module->SetDistance(distance_to_goal));
+        OKC_CALL(this->right_back_module->SetDistance(distance_to_goal));
+
+        OKC_CALL(this->left_front_module->GetDriveOutput(&this->interface_->left_front_drive_motor_output));
+        OKC_CALL(this->left_back_module->GetDriveOutput(&this->interface_->left_back_drive_motor_output));
+        OKC_CALL(this->right_front_module->GetDriveOutput(&this->interface_->right_front_drive_motor_output));
+        OKC_CALL(this->right_back_module->GetDriveOutput(&this->interface_->right_back_drive_motor_output));
+    } else {
+        // we shouldn't have gotten here, throw an error
+        return false;
+    }
+
+    //TODO do I need to call Update() on the modules? periodic should still technically get called, I think
+
+    // set motor outputs
+
+    OKC_CALL(this->left_front_module->GetSteerOutput(&this->interface_->left_front_steer_motor_output));
+    OKC_CALL(this->left_back_module->GetSteerOutput(&this->interface_->left_back_steer_motor_output));
+    OKC_CALL(this->right_front_module->GetSteerOutput(&this->interface_->right_front_steer_motor_output));
+    OKC_CALL(this->right_back_module->GetSteerOutput(&this->interface_->right_back_steer_motor_output));
+
+    return true;
+}
+
+bool SwerveDrive::GetHeading(double *heading) {
+    OKC_CHECK(heading != nullptr);
+    OKC_CHECK(interface_ != nullptr);
+
+    heading = &interface_->imu_yaw;
+
+    return true;
+}
+
+bool SwerveDrive::GetLeftDriveEncoderAverage(double *avg) {
+    OKC_CHECK(interface_ != nullptr);
+
+    double tmp = 0;
+
+    tmp += interface_->left_front_drive_motor_enc;
+    tmp += interface_->left_back_drive_motor_enc;
+
+    *avg = (tmp / 2);
+
+    return true;
+}
+
+bool SwerveDrive::GetRightDriveEncoderAverage(double *avg) {
+    OKC_CHECK(interface_ != nullptr);
+
+    double tmp = 0;
+    
+    tmp += interface_->right_front_drive_motor_enc;
+    tmp += interface_->right_back_drive_motor_enc;
+
+    *avg = (tmp / 2);
+
+    return true;
+}
+
+bool SwerveDrive::GetDriveEncoderAverage(double *avg) {
```

Comment:
Are these functions useful for swerve?

Reply:
I kinda just had them there on shuffleboard to make sure things were actually working. They're somewhat useful for swerve, but individual NetworkTableEntries for each module would be best